run/O2SimDeviceRunner.cxx

```diff
@@ -23,14 +23,36 @@
 #include <sys/wait.h>
 #include <pthread.h> // to set cpu affinity
 #include <cmath>
+#include <signal.h>
+#include <unistd.h>
 
 #include "rapidjson/document.h"
 #include "rapidjson/writer.h"
 #include "rapidjson/stringbuffer.h"
 #include "rapidjson/filereadstream.h"
-
 namespace bpo = boost::program_options;
 
+std::vector<int> gChildProcesses; // global vector of child pids
+int gMasterProcess = -1;
+
+// custom signal handler to ensure that we
+// distribute signals to all detached forks
+void sighandler(int signal)
+{
+  if (signal == SIGTERM || signal == SIGINT) {
+    auto pid = getpid();
+    if (pid == gMasterProcess) {
+      // the master
+      for (auto child : gChildProcesses) {
+        kill(child, signal);
+      }
+      // wait for all children in the same process group
+      waitpid(0, nullptr, WNOHANG);
```

Comment:
I would do the waiting in the SIGCHLD handler. This will still create zombies in the case the wait happens before the signal is notified to the child (there is no guarantee on the synchronicity of signals).